### ISSUE TITLE
Support for byte arrays in class providers

### ISSUE THREAD

**Issue by SoniaZaldana** [soot-oss/SootUp#621]
Hi team, 

I’m working on a constant propagation tool to conduct some optimizations at [link time](https://openjdk.org/jeps/261). I’m running into some limitations with SootUp with regards to specifying custom input locations for class analysis. 

As of JDK9+, all of Java’s standard libraries have been broken up into [ JMODs](https://docs.oracle.com/en/java/javase/11/tools/jmod.html), which are then included along with explicit and transitive JARs into a minified JREs. 

To conduct our optimizations, we need to be able to parse Java’s standard libraries available in the JMODs at link time. In our case, JLink provisions class file contents as a collection of byte[] entries. We currently do a lot of work with ASM at link time passing on the class file contents to ASM class readers as [byte arrays](https://asm.ow2.io/javadoc/org/objectweb/asm/ClassReader.html#%3Cinit%3E(byte%5B%5D)).

It is my understanding that the APIs for generating the class sources in SootUp are reliant on a path. Note, for example, the contract for [ClassProvider](https://github.com/soot-oss/SootUp/blob/develop/sootup.core/src/main/java/sootup/core/frontend/ClassProvider.java) used in the [AsmJavaClassProvider](https://github.com/soot-oss/SootUp/blob/develop/sootup.java.bytecode/src/main/java/sootup/java/bytecode/frontend/AsmJavaClassProvider.java). 

We have currently made changes to SootUp to support input locations denoting class file contents as byte arrays, but we feel there’s a strong case to come up with a more robust solution upstream. 

Would it be possible to do so? 

Thank you in advance for your help!

Sonia

**Comment by swissiety** [comment]
Hi Sonia,
thanks for your contributing your thoughts!

Basically the idea was to create a resolving mechanism to load inputs "automatically" i.e. via a mapping of:
file-suffix -> specific inputlocation 
So far so good - obviously that didnt happen but the ClassProvider structure survived.

As far as i can see the `ClassProvider` interface adds no real value in the current state/architecture
(Each AnalysisInputLocation just keeps a reference to 'its' provider and its only used there). So removing the `ClassProvider` interface would be fine and allows for other ways of refering to the compilation unit (contents) like in the situation you described.

Markus

**Comment by SoniaZaldana** [comment]
Hi Markus, 

Thanks for your response! I think you're right and I have been thinking about ways to incorporate the byte array support with the existing APIs. I did have a follow up question for you. 

I’m trying to write my own class provider without making use of the ```ClassProvider``` interface as you described. However, I noticed that the classes responsible for reading java bytecode (namely AsmMethodSource and AsmClassSource) are both restricted to the package. Ideally, I'd like to recycle these to write a sensible class provider using byte arrays, as I'm not interested in changing or maintaining the logic on how to interpret the java bytecode. 

I wonder what your thoughts are on opening those up? 

Thanks in advance! 

Sonia
